File: mpi.c
#include <stddef.h>
#include <mpi.h>

#ifdef WINDOWS
#define DLLEXPORT __declspec(dllexport)
#else
#define DLLEXPORT 
#endif

#ifdef MSMPI
#define MpiType long long 
#endif

#ifdef OpenMPI
#define MpiType void*
#endif

#ifdef MPICH
#define MpiType long long
#endif

typedef struct { double re; double im; } complex16;

DLLEXPORT MPI_Op GetMpiOpSum() { return MPI_SUM; }

DLLEXPORT MpiType GetCommWorld() { return MPI_COMM_WORLD; }

DLLEXPORT MpiType GetMpiInt() { return MPI_INT; }
DLLEXPORT MpiType GetMpiFloat() { return MPI_FLOAT; }
DLLEXPORT MpiType GetMpiDouble() { return MPI_DOUBLE; }
DLLEXPORT MpiType GetMpiDoubleComplex() { return MPI_C_DOUBLE_COMPLEX; }


DLLEXPORT int GetMaxProcessorName() { return MPI_MAX_PROCESSOR_NAME; }
DLLEXPORT int GetMpiAnySource() { return MPI_ANY_SOURCE; }


DLLEXPORT int GetErrorString(int errorcode, char *string, int *resultlen)
{
	return MPI_Error_string(errorcode, string, resultlen);
}

DLLEXPORT int Init()
{
	return MPI_Init(0, NULL);
}

DLLEXPORT int Barrier(MPI_Comm comm)
{
	return MPI_Barrier(comm);
}

DLLEXPORT int AllGatherV(void* sendbuf, int size, void* rbuf, int *recvcounts, int *displs)
{
	return MPI_Allgatherv(sendbuf, size, MPI_DOUBLE_COMPLEX, rbuf, recvcounts, displs, MPI_DOUBLE_COMPLEX, MPI_COMM_WORLD);
}

DLLEXPORT int Gather(void *sendbuf, int sendcount, void *recvbuf, int recvcount, int root, MPI_Comm comm)
{
	return MPI_Gather(sendbuf, sendcount, MPI_DOUBLE_COMPLEX, recvbuf, recvcount, MPI_DOUBLE_COMPLEX, root, comm);
}

DLLEXPORT int GatherV(void* sendbuf, int size, void* rbuf, int *recvcounts, int *displs)
{
	return MPI_Gatherv(sendbuf, size, MPI_DOUBLE_COMPLEX, rbuf, recvcounts, displs, MPI_DOUBLE_COMPLEX, 0, MPI_COMM_WORLD);
}

DLLEXPORT int AllReduce(void *sendbuf, void *recvbuf, int count, MPI_Datatype datatype, MPI_Comm comm)
{
	return MPI_Allreduce(sendbuf, recvbuf, count, datatype, MPI_SUM, comm);
}

DLLEXPORT int Reduce(void *sendbuf, void *recvbuf, int count, MPI_Datatype datatype, MPI_Comm comm)
{
	return MPI_Reduce(sendbuf, recvbuf, count, datatype, MPI_SUM, 0, comm);
}

DLLEXPORT int GetProcessorName(char *name, int *resultlen)
{
	return MPI_Get_processor_name(name, resultlen);
}

DLLEXPORT int GetCommWorldRank(int* rank)
{
	return MPI_Comm_rank(MPI_COMM_WORLD, rank);
}

DLLEXPORT int GetCommWorldSize(int* size)
{
	return MPI_Comm_size(MPI_COMM_WORLD, size);
}

DLLEXPORT int CommCreate(MPI_Comm comm, MPI_Group group, MPI_Comm *newcomm)
{
	return MPI_Comm_create(comm, group, newcomm);
}

DLLEXPORT int GroupIncl(MPI_Group group, int n, int ranks[], MPI_Group *newgroup)
{
	return MPI_Group_incl(group, n, ranks, newgroup);
}

DLLEXPORT int CommGroup(MPI_Comm comm, MPI_Group *group)
{
	return MPI_Comm_group(comm, group);
}

DLLEXPORT int Bcast(void *buffer, int count, MPI_Datatype datatype, int root, MPI_Comm comm)
{
	return	MPI_Bcast(buffer, count, datatype, root, comm);
}

DLLEXPORT int Send(void* data, int count, MPI_Datatype datatype, int dest, int tag, MPI_Comm comm)
{
	return MPI_Send(data, count, datatype, dest, tag, comm);
}

DLLEXPORT int Recv(void *buf, int count, MPI_Datatype datatype, int source, int tag, MPI_Comm comm, int* actualSource)
{
	MPI_Status status;
	int err = MPI_Recv(buf, count, datatype, source, tag, comm, &status);
	(*actualSource) = status.MPI_SOURCE;
	return err;
}

DLLEXPORT int SendComplexMatrix(void* data, int nx, int ny, int destination, int tag)
{
	return MPI_Send(data, nx*ny, MPI_DOUBLE_COMPLEX, destination, tag, MPI_COMM_WORLD);
}

DLLEXPORT int AllToAllDoubleComplex(complex16 *sendbuf, int sendcount, void *recvbuf, int recvcount, MPI_Comm comm)
{
	return MPI_Alltoall(sendbuf, sendcount, MPI_DOUBLE_COMPLEX, recvbuf, recvcount, MPI_DOUBLE_COMPLEX, comm);
}

DLLEXPORT int AllToAllDoubleComplexInPlace(void *recvbuf, int recvcount, MPI_Comm comm)
{
	return MPI_Alltoall(MPI_IN_PLACE, recvcount, MPI_DOUBLE_COMPLEX, recvbuf, recvcount, MPI_DOUBLE_COMPLEX, comm);
}

DLLEXPORT void Finalize()
{
	MPI_Finalize();
}

File: Extreme.Mpi/Mpi/Mpi.cs
using System;
using System.Numerics;
using System.Text;

using static Extreme.Parallel.UnsafeNativeMethods;
using UNM = Extreme.Parallel.UnsafeNativeMethods;
using System.ComponentModel.Design;
using System.Runtime.Remoting.Messaging;
using System.Runtime.ConstrainedExecution;

namespace Extreme.Parallel
{
    public unsafe class Mpi : IDisposable
    {
        public readonly int Master = 0;

		private static readonly IntPtr CommWorld; //= GetCommWorld();

		private static readonly IntPtr CommNull;// =GetCommNull();

		public static readonly IntPtr Int; 
		public static readonly IntPtr Float; 
		public static readonly IntPtr Double;
		public static readonly IntPtr Complex;
		public static readonly int AnySource; 
		public static readonly IntPtr InPlace;
		public static readonly IntPtr MpiOpSumm;
		private static readonly bool _mpiexists=true;

        private readonly int _rank = 0;
        private readonly int _size = 1;
        private Mpi() {
			
			IntPtr comm;
			if (_mpiexists) {
				Communicator = CommWorld;

				int rank;
				int size;

				UnsafeNativeMethods.GetCommRank (Communicator, &rank);
				UnsafeNativeMethods.GetCommSize (Communicator, &size);
				_rank = rank;
				_size = size;
			} else {
				_rank = 0;
				_size = 1;
			}
		}



        private static bool _initialized = false;

		static Mpi(){
			try{
				UnsafeNativeMethods.Init();
				CommWorld = GetCommWorld();
				CommNull =GetCommNull();
				Int = GetMpiInt();
				Float = GetMpiFloat();
				Double = GetMpiDouble();
				Complex = GetMpiDoubleComplex();
				AnySource = GetMpiAnySource();
				InPlace=GetMpiInPlace();
				MpiOpSumm = GetMpiOpSum();
				_initialized = true;
			}catch (DllNotFoundException){
				_mpiexists=false;
			}
		}


        public static Mpi Init()
        {
			
            try
            {
				return new Mpi();
			}
            finally
            {
                _initialized = true;
            }
        }
		private Mpi(IntPtr comm)
		{
			if (_initialized) {
				Communicator = comm;

				int rank;
				int size;
				int err;
				err = UnsafeNativeMethods.GetCommRank (comm, &rank);
					
				if (err != 0)
					throw new InvalidOperationException (GetErrorString (err));
				_rank = rank;

				err = UnsafeNativeMethods.GetCommSize (comm, &size);
				if (err != 0)
					throw new InvalidOperationException (GetErrorString (err));

				_size = size;
			} else {
				
			}
		}
		public Mpi Merge (params int[] ranks)
		{
			var comm=CreateNewCommunicator(ranks);
			if (comm != CommNull)
				return new Mpi (comm);
			else
				return null;
		}

		public Mpi Separate(int color){
			IntPtr comm;
			UNM.CommSplit (Communicator, color, Rank, &comm);
			return new Mpi (comm);
		}


		public bool IsParallel => Size > 1;
        public int Size => _size;
        public int Rank => _rank;
        public bool IsMaster => _rank == Master;
		public IntPtr Communicator { get; private set; }=CommNull;

        private static void WithErrorHandling(int err)
        {
            if (err != 0)
                throw new InvalidOperationException(GetErrorString(err));
        }

        private IntPtr CreateNewCommunicator(params int[] ranks)
        {
            IntPtr group;
            CommGroup(Communicator, &@group);

            IntPtr newGroup;
            GroupIncl(@group, ranks.Length, ranks, &newGroup);

            IntPtr newComm;
			CommCreate(Communicator, newGroup, &newComm);

            return newComm;
        }


		public Mpi Dup(){
			IntPtr comm;
			if (_mpiexists) {
				UNM.CommDup (this.Communicator, &comm);
				return new Mpi (comm);
			} else {
				return new Mpi();
			}
		}

        public Complex AllReduce(Complex value)
        {
            Complex result;
			if (_mpiexists) {
				UnsafeNativeMethods.AllReduce(&value, &result, 2, Double, Communicator);
			} else {
				result = value;
			}
           	return result;
        }

		public void AllReduceInPlace(Complex[] values)
		{
			fixed(Complex* ptr=&values[0]){
				UnsafeNativeMethods.AllReduceInPlace(ptr, 2*values.Length, Double, Communicator);
			}


		}



        public double AllReduce(double value)
        {
            double result;
			if (_mpiexists) {
				UnsafeNativeMethods.AllReduce (&value, &result, 1, Double, Communicator);
			} else {
				result = value;
			}
            return result;
        }

		public int AllReduce(int value)
		{
			int result;
			Console.WriteLine("int");
			if (_mpiexists) {
				UnsafeNativeMethods.AllReduce(&value, &result, 1, Int, Communicator);
			} else {
				result = value;
			}
			return result;
		}


        public void Reduce(double* value, double* result, int length)
        { 
			if (_mpiexists) {
				if (result == value && IsMaster) {
					UnsafeNativeMethods.Reduce ((void*)InPlace, result, length, Double, Communicator);
				} else {
					UnsafeNativeMethods.Reduce (value, result, length, Double, Communicator);
				}
			} else {
				if (result == value) {
					return;
				} else {
					for (int i = 0; i < length; i++) {
						result [i] = value [i];
					}
				}
			}
        }
		public void Reduce(double[] value, double[] result)
		{
			if (_mpiexists) {
				fixed(double* pv=&value[0], ptr=&result[0]) {
					if (pv == ptr && IsMaster) {
						UnsafeNativeMethods.Reduce ((void*)InPlace, ptr, value.Length, Double, Communicator);
					} else {
						UnsafeNativeMethods.Reduce (pv, ptr, value.Length, Double, Communicator);
					}
				}
			} else {
				if (result == value) {
					return;
				} else {
					for (int i = 0; i < result.Length; i++) {
						result [i] = value [i];
					}
				}
			}
		}


		public double Reduce(double value)
		{
			double result = value;
			if (_mpiexists)
				UnsafeNativeMethods.Reduce (&value, &result,1, Double, Communicator);
			return result;
		}





		public void LogicalReduce(double* value, double* result, int length)
		{

			if (result == value && IsMaster) {
				UnsafeNativeMethods.LogicalReduce ((void*)InPlace, result, length,  Communicator);
			} else {
				UnsafeNativeMethods.LogicalReduce (value, result, length,  Communicator);
			}
		}

		public static long CommunicatorC2Fortran(IntPtr comm)
		{
			return UnsafeNativeMethods.CommunicatorC2Fortran(comm);
		}

		public long CommunicatorC2Fortran()
		{
			return CommunicatorC2Fortran(this.Communicator);
		}

		public static void Barrier(IntPtr comm)
            => WithErrorHandling(UnsafeNativeMethods.Barrier(comm));

		public void Barrier()
		=> WithErrorHandling(UnsafeNativeMethods.Barrier(Communicator));

        public int BroadCast(int root, int value=0)
        {
			var tmp = value;
			Bcast(&tmp, 1, Int, root, Communicator);
           return tmp;
        }






        public void AllGatherV(Complex* src, int sendSize, Complex* dst, int[] rCounts, int[] rDispl)
        {
            fixed (int* cntPtr = &rCounts[0], dsplPtr = &rDispl[0])
			UnsafeNativeMethods.AllGatherV(src, sendSize, dst, cntPtr, dsplPtr,Communicator);
        }

        public void GatherV(Complex* src, int sendSize, Complex* dst, int[] rCounts, int[] rDispl)
        {
            fixed (int* cntPtr = &rCounts[0], dsplPtr = &rDispl[0])
			UnsafeNativeMethods.GatherV(src, sendSize, dst, cntPtr, dsplPtr,Communicator);
        }

        public void BroadCast( int root, double[] values)
        {
            fixed (double* ptr = &values[0])
			Bcast(ptr, values.Length, Double, root, Communicator);
        }

		public void BroadCast( int root, int[,,] values)
		{
			fixed (int* ptr = &values[0,0,0])
				BroadCast ( root, ptr, values.Length);
		}


        public void BroadCast(int root, Complex[] values)
        {
            fixed (Complex* ptr = &values[0])
			Bcast(ptr, values.Length, Complex, root, Communicator);
        }

        public void BroadCast(int root, double* values, int length)
        {
            Bcast(values, length, Double, root, Communicator);
        }


        public void BroadCast(int root, Complex* values, int length)
        {
			Bcast(values, length, Complex, root, Communicator);
        }

		public void BroadCast( int root, int* values, int length)
		{
			Bcast(values, length, Int, root, Communicator);
		}

		public int MasterBroadCast(int value=0)
		{
			var tmp = value;
			if (_mpiexists)
				Bcast(&tmp, 1, Int, Master, Communicator);
			return tmp;
		}



		public void MasterBroadCast(int[,] data)
		{
			var len = data.Length;
			fixed(int* ptr=&data[0,0]) {
				Bcast (ptr, len, Int, Master, Communicator);
			}

		}

		public void MasterBroadCast(double[,] data)
		{
			var len = data.Length;
			fixed(double* ptr=&data[0,0]) {
				Bcast (ptr, len, Double, Master, Communicator);
			}

		}


        public int Send(void* data, int count, IntPtr datatype, int dest, int tag)
		=> UNM.Send(data, count, datatype, dest, tag, Communicator);

        public int Recv(void* data, int count, IntPtr datatype, int source, int tag, out int actualSource)
		=> UNM.Recv(data, count, datatype, source, tag, Communicator, out actualSource);

        public static string GetErrorString(int error)
            => UNM.GetErrorString(error);

        public string GetProcessorName()
        {
            var name = new byte[GetMaxProcessorName()];
            int length = 0;

            fixed (byte* namePtr = &name[0])
            {
                int err = UNM.GetProcessorName(namePtr, ref length);
            }

            return Encoding.ASCII.GetString(name, 0, length);
        }

	
        public void Gather(Complex* dst, Complex* src, int dstSize, int srcSize)
        {
			if (_mpiexists) {
				UNM.Gather (src, srcSize, dst, dstSize, Master, Communicator);
			} else {
				if (src != dst) {
					for (int i = 0; i < srcSize; i++)
						dst [i] = src [i];
				}
			}
        }

        public void Gather(Complex[] dst, Complex src)
        {
			if (_mpiexists){
            	fixed (Complex* dstPtr = &dst[0])
					UNM.Gather(&src, 1, dstPtr, 1, Master, Communicator);
			} else {
				dst [0] = src;
			}
		}

        public void AllToAll(Complex* buffer, int size)
		=> AllToAllDoubleComplexInPlace(buffer, size, Communicator);

        public void AllToAll(Complex* src, int srcSize, Complex* dst, int dstSize)
		=> AllToAllDoubleComplex(src, srcSize, dst, dstSize, Communicator);


		public void Free(){
			IntPtr comm = Communicator;
			if (Communicator!=CommNull&&Communicator!=CommWorld)
				UnsafeNativeMethods.CommFree(&comm);
			Communicator = CommNull;
		}

        public void Dispose()
        {
           // FinalizeMpi();
			if (_mpiexists) {
				if (Communicator == CommWorld)
					Finalize ();
					Free ();
			}
        }

		private static void Finalize()
		{
			if (_mpiexists)
				FinalizeMpi ();
		}
    }
}
